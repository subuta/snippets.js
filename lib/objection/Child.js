import { build, format, snippets as s } from 'bld.js'
import _ from 'lodash'
import pluralize from 'pluralize'

export default ({model, config}, children = []) => {
  const {
    schema = {}
  } = config

  const Model = _.upperFirst(pluralize.singular(_.toLower(model)))

  const imports = s.import([
    ['./Model', 'Model'],
    ['lodash', '_']
  ])

  const parseSchema = (schema) => {
    // Expects schema has following keys.
    // tableName = will transform to Model's tableName
    // required = will transform to Model's jsonSchema
    // properties = will transform to Model's jsonSchema
    // relations = will transform to Model's relationMappings
    let {
      tableName,
      required,
      properties,
      relations
    } = schema

    // ensure naming convention.
    const model = pluralize.singular(_.toLower(tableName))
    const models = pluralize(model)
    const Model = _.upperFirst(_.camelCase(model))

    const parseRelation = (relation) => _.transform(relation, (results, value, key) => {
      if (key === 'morphAs') {
        const morphType = s.camelCase(`${value}Type`)
        results['filter'] = {
          [morphType]: s.stringify(Model)
        }
        results['beforeInsert'] = build`
          (model) => {
            model.${morphType} = ${s.stringify(Model)}
          }
        `
        return undefined
      }

      // pass-through
      if (key === 'join') {
        results[key] = s.stringify(value)
        return undefined
      }

      const relatedModel = pluralize.singular(_.toLower(value))
      const RelatedModel = _.upperFirst(relatedModel)

      results['modelClass'] = `models.${RelatedModel}`
      switch (key) {
        case 'belongsTo':
          results['relation'] = 'Model.BelongsToOneRelation'
          break
        case 'hasOne':
          results['relation'] = 'Model.HasOneRelation'
          break
        case 'hasMany':
          results['relation'] = 'Model.HasManyRelation'
          break
        case 'hasAndBelongsToMany':
          results['relation'] = 'Model.ManyToManyRelation'
          break
      }
    }, {})

    const relationMappings = _.transform(relations, (results, value, key) => {
      results[key] = parseRelation(value)
    }, {})

    const jsonSchema = {
      title: s.stringify(Model),
      $id: s.stringify(`http://sub-labo.com/schemas/${model}.json`),
      type: s.stringify('object'),
      required: s.stringify(required),
      properties: s.stringify(properties)
    }

    return {
      tableName: models,
      relationMappings,
      jsonSchema
    }
  }

  let {
    tableName,
    relationMappings,
    jsonSchema
  } = parseSchema(schema)

  let {
    properties,
    visible,
    hidden
  } = schema

  const booleanProps = _.map(_.pickBy(properties, {type: 'boolean'}), (v, k) => {
    return `
      if (_.isNumber(parseInt(json['${k}']))) {
        json['${k}'] = !!json['${k}']
      }
    ` + s.EOL
  })

  return build`
    ${imports}
    
   
    ${s.export(build`
      class ${Model} extends Model {
        // SEE: https://github.com/Vincit/objection.js/issues/825
        $parseDatabaseJson(db) {
          // Remember to call the super class's implementation.
          const json = super.$parseDatabaseJson(db);
          ${booleanProps}
          return json
        }
        
        $formatDatabaseJson(json) {
          return super.$formatDatabaseJson(json);
        }
      
        static register = (models) => {
          // then define relationMappings.
          ${Model}.relationMappings = ${s.raw(relationMappings)}
        }
      
        static tableName = '${tableName}'
        
        static jsonSchema = ${s.raw(jsonSchema)}
        
        ${visible && build`
          static get visible() {
            return ${s.stringify(visible)}
          }
        `}
        
        ${hidden && build`
          static get hidden() {
            return ${s.stringify(hidden)}
          }
        `}
        
        /* mat Custom methods [start] */
        /* mat Custom methods [end] */
      }
    `)}
  `
}
