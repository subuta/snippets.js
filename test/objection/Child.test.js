/* global expect, describe, it */

import { build, format, snippets as s } from 'bld.js'

import Child from 'lib/objection/Child'

describe('objection/Child', () => {
  it('should create objection Child', () => {
    const Book = {
      tableName: 'books',
      required: [
        'title',
      ],

      properties: {
        id: {
          type: 'integer'
        },
        title: {
          'type': 'string'
        }
      },

      relations: {
        comments: {
          hasMany: 'comments',
          join: {
            from: 'books.id',
            to: 'comments.bookId'
          }
        },
      }
    }

    expect(format(Child({
      model: 'Book',
      config: {
        schema: Book
      }
    }))).toMatchSnapshot()
  })

  it('should create objection Child with Boolean property', () => {
    const Book = {
      tableName: 'books',
      required: [
        'title',
      ],

      properties: {
        id: {
          type: 'integer'
        },
        title: {
          'type': 'string'
        },
        isOutOfPrint: {
          'type': 'boolean'
        }
      },

      relations: {
        comments: {
          hasMany: 'comments',
          join: {
            from: 'books.id',
            to: 'comments.bookId'
          }
        },
      }
    }

    expect(format(Child({
      model: 'Book',
      config: {
        schema: Book
      }
    }))).toMatchSnapshot()
  })
})
