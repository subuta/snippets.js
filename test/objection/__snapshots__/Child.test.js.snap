// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`objection/Child should create objection Child 1`] = `
"import Model from './Model'
import _ from 'lodash'

export default class Book extends Model {
  // SEE: https://github.com/Vincit/objection.js/issues/825
  $parseDatabaseJson(db) {
    // Remember to call the super class's implementation.
    const json = super.$parseDatabaseJson(db)

    return json
  }

  $formatDatabaseJson(json) {
    return super.$formatDatabaseJson(json)
  }

  static register = (models) => {
    // then define relationMappings.
    Book.relationMappings = {
      comments: {
        modelClass: models.Comment,
        relation: Model.HasManyRelation,
        join: {from: 'books.id', to: 'comments.bookId'}
      }
    }
  }

  static tableName = 'books'

  static jsonSchema = {
    title: 'Book',
    $id: 'http://sub-labo.com/schemas/book.json',
    type: 'object',
    required: ['title'],
    properties: {id: {type: 'integer'}, title: {type: 'string'}}
  }
}
"
`;

exports[`objection/Child should create objection Child with Boolean property 1`] = `
"import Model from './Model'
import _ from 'lodash'

export default class Book extends Model {
  // SEE: https://github.com/Vincit/objection.js/issues/825
  $parseDatabaseJson(db) {
    // Remember to call the super class's implementation.
    const json = super.$parseDatabaseJson(db)

    if (_.isNumber(parseInt(json['isOutOfPrint']))) {
      json['isOutOfPrint'] = !!json['isOutOfPrint']
    }

    return json
  }

  $formatDatabaseJson(json) {
    return super.$formatDatabaseJson(json)
  }

  static register = (models) => {
    // then define relationMappings.
    Book.relationMappings = {
      comments: {
        modelClass: models.Comment,
        relation: Model.HasManyRelation,
        join: {from: 'books.id', to: 'comments.bookId'}
      }
    }
  }

  static tableName = 'books'

  static jsonSchema = {
    title: 'Book',
    $id: 'http://sub-labo.com/schemas/book.json',
    type: 'object',
    required: ['title'],
    properties: {
      id: {type: 'integer'},
      title: {type: 'string'},
      isOutOfPrint: {type: 'boolean'}
    }
  }
}
"
`;

exports[`objection/Child should create objection Child with Boolean property 2`] = `
"import Model from './Model'
import _ from 'lodash'

export default class Book extends Model {
  // SEE: https://github.com/Vincit/objection.js/issues/825
  $parseDatabaseJson(db) {
    // Remember to call the super class's implementation.
    const json = super.$parseDatabaseJson(db)

    if (_.isNumber(parseInt(json['isOutOfPrint']))) {
      json['isOutOfPrint'] = !!json['isOutOfPrint']
    }

    return json
  }

  $formatDatabaseJson(json) {
    return super.$formatDatabaseJson(json)
  }

  static register = (models) => {
    // then define relationMappings.
    Book.relationMappings = {
      comments: {
        modelClass: models.Comment,
        relation: Model.HasManyRelation,
        join: {from: 'books.id', to: 'comments.bookId'}
      }
    }
  }

  static tableName = 'books'

  static jsonSchema = {
    title: 'Book',
    $id: 'http://sub-labo.com/schemas/book.json',
    type: 'object',
    required: ['title'],
    properties: {
      id: {type: 'integer'},
      title: {type: 'string'},
      isOutOfPrint: {type: 'boolean'}
    }
  }
}
"
`;

exports[`objection/Child should create objection Child with MorphAs property 1`] = `
"import Model from './Model'
import _ from 'lodash'

export default class Book extends Model {
  // SEE: https://github.com/Vincit/objection.js/issues/825
  $parseDatabaseJson(db) {
    // Remember to call the super class's implementation.
    const json = super.$parseDatabaseJson(db)

    if (_.isNumber(parseInt(json['isOutOfPrint']))) {
      json['isOutOfPrint'] = !!json['isOutOfPrint']
    }

    return json
  }

  $formatDatabaseJson(json) {
    return super.$formatDatabaseJson(json)
  }

  static register = (models) => {
    // then define relationMappings.
    Book.relationMappings = {
      comments: {
        modelClass: models.Comment,
        relation: Model.HasManyRelation,
        filter: {
          commentableType: 'Book'
        },
        beforeInsert: (model) => {
          model.commentableType = 'Book'
        },
        join: {from: 'books.id', to: 'comments.commentableId'}
      }
    }
  }

  static tableName = 'books'

  static jsonSchema = {
    title: 'Book',
    $id: 'http://sub-labo.com/schemas/book.json',
    type: 'object',
    required: ['title'],
    properties: {
      id: {type: 'integer'},
      title: {type: 'string'},
      isOutOfPrint: {type: 'boolean'}
    }
  }
}
"
`;
